File: typescript/bots/simple-oracle-limit-maker/src/maker.ts
import {
	DriftClient,
	DLOBSubscriber,
	OrderSubscriber,
	EventSubscriber,
	initialize,
	convertToNumber,
	MarketType,
	OrderType,
	PositionDirection,
	BASE_PRECISION,
	PerpMarkets,
	Wallet,
	type OraclePriceData,
	type PerpMarketConfig,
	SlotSubscriber,
	PostOnlyParams,
	QUOTE_PRECISION,
	BN,
	loadKeypair,
} from '@drift-labs/sdk';
import { Connection, Keypair, PublicKey } from '@solana/web3.js';
import { makeLogger } from './utils';
import { MM_CONFIG } from './config';

const log = makeLogger('example-maker');

// Bot state tracking
type BotState = {
	oraclePrice: number;
	previousOraclePrice: number;
	lastOracleUpdate: number;
	hasActiveOrders: boolean;
	totalMatches: number;
	realizedPnl: number;
	isRunning: boolean;
	isProcessing: boolean;
};

// FIFO queue entry for P&L tracking
type FillEntry = {
	timestamp: number;
	size: number;
	price: number;
};

/**
 * Simple market maker that provides liquidity by placing bid/ask orders
 * Core loop: listen to oracle → calculate quotes → place orders → track fills
 */
export class ExampleMaker {
	// Drift protocol components
	private driftClient!: DriftClient;
	private slotSubscriber!: SlotSubscriber;
	private dlobSubscriber!: DLOBSubscriber;
	private orderSubscriber!: OrderSubscriber;
	private eventSubscriber!: EventSubscriber;

	// Market configuration
	private marketIndex!: number;
	private marketConfig!: PerpMarketConfig;

	// Bot state
	private state: BotState = {
		oraclePrice: 0,
		previousOraclePrice: 0,
		lastOracleUpdate: 0,
		hasActiveOrders: false,
		totalMatches: 0,
		realizedPnl: 0,
		isRunning: false,
		isProcessing: false,
	};

	// FIFO queues for P&L calculation
	private buyQueue: FillEntry[] = [];
	private sellQueue: FillEntry[] = [];

	constructor() {
		// Find market configuration
		const market = PerpMarkets[MM_CONFIG.ENV].find(
			(m) => m.baseAssetSymbol === MM_CONFIG.TARGET_MARKET
		);
		if (!market) {
			throw new Error(`Market ${MM_CONFIG.TARGET_MARKET} not found`);
		}

		this.marketIndex = market.marketIndex;
		this.marketConfig = market;

		log.info('MARKET', 'Market found', {
			symbol: MM_CONFIG.TARGET_MARKET,
			marketIndex: this.marketIndex,
		});
	}

	async initialize(): Promise<{ success: boolean; error?: string }> {
		try {
			log.info('INIT', 'Initializing Example Maker');

			const initResult = await this.initializeDrift();
			if (!initResult.success) return initResult;

			await this.initializeSubscribers();
			this.setupOracleListener();
			await this.setupFillEventListener();

			log.info('INIT', 'Example Maker initialized successfully');
			return { success: true };
		} catch (error) {
			log.error('INIT', 'Initialization failed', error as Error);
			return {
				success: false,
				error: `Initialization failed: ${(error as Error).message}`,
			};
		}
	}

	/**
	 * Initialize Drift client and authenticate wallet
	 */
	private async initializeDrift(): Promise<{
		success: boolean;
		error?: string;
	}> {
		try {
			const sdk = initialize({ env: MM_CONFIG.ENV });
			const connection = new Connection(MM_CONFIG.RPC_ENDPOINT, {
				commitment: 'confirmed',
			});

			let wallet: Wallet;

			if (MM_CONFIG.KEYPAIR_PATH) {
				// Use keypair file if path is provided
				wallet = new Wallet(loadKeypair(MM_CONFIG.KEYPAIR_PATH));
			} else if (MM_CONFIG.PRIVATE_KEY) {
				// Fall back to private key string
				const parseKey = (key: string): Uint8Array =>
					key.startsWith('[')
						? new Uint8Array(JSON.parse(key))
						: Buffer.from(key, 'base64');

				const secretKey = parseKey(MM_CONFIG.PRIVATE_KEY);
				wallet = new Wallet(Keypair.fromSecretKey(secretKey));
			} else {
				throw new Error(
					'Either KEYPAIR_PATH or PRIVATE_KEY environment variable must be set'
				);
			}

			// Initialize Drift client
			this.driftClient = new DriftClient({
				connection,
				wallet,
				env: MM_CONFIG.ENV,
				programID: new PublicKey(sdk.DRIFT_PROGRAM_ID),
				accountSubscription: {
					type: 'websocket',
				},
			});

			await this.driftClient.subscribe();
			await this.driftClient.getUser().exists();

			log.info('DRIFT', 'Drift client initialized successfully', {
				publicKey: wallet.publicKey.toString(),
				env: MM_CONFIG.ENV,
			});

			return { success: true };
		} catch (error) {
			return {
				success: false,
				error: `Drift initialization failed: ${(error as Error).message}`,
			};
		}
	}

	/**
	 * Initialize data subscribers for market data and events
	 */
	private async initializeSubscribers(): Promise<void> {
		try {
			log.info('SUBSCRIBERS', 'Initializing subscribers');

			// Slot subscriber for blockchain state
			this.slotSubscriber = new SlotSubscriber(this.driftClient.connection);

			// Order book subscriber
			this.orderSubscriber = new OrderSubscriber({
				driftClient: this.driftClient,
				subscriptionConfig: {
					type: 'websocket',
					commitment: 'processed',
					resubTimeoutMs: 30000,
				},
			});

			// DLOB (Decentralized Limit Order Book) subscriber
			this.dlobSubscriber = new DLOBSubscriber({
				dlobSource: this.orderSubscriber,
				slotSource: this.slotSubscriber,
				driftClient: this.driftClient,
				updateFrequency: 1000,
			});

			// Event subscriber for fill notifications
			this.eventSubscriber = new EventSubscriber(
				this.driftClient.connection,
				this.driftClient.program,
				{
					eventTypes: ['OrderActionRecord'],
					commitment: 'confirmed',
					logProviderConfig: {
						type: 'websocket',
					},
				}
			);

			// Start all subscriptions
			await this.orderSubscriber.subscribe();
			await this.dlobSubscriber.subscribe();
			await this.eventSubscriber.subscribe();
			await this.slotSubscriber.subscribe();

			log.info('SUBSCRIBERS', 'All subscribers initialized successfully');
		} catch (error) {
			throw new Error(
				`Subscriber initialization failed: ${(error as Error).message}`
			);
		}
	}

	/**
	 * Setup oracle price feed listener
	 * This is the primary trigger for quote updates
	 */
	private setupOracleListener(): void {
		log.info('ORACLE', 'Setting up oracle listener');

		this.driftClient.eventEmitter.on(
			'oraclePriceUpdate',
			(publicKey: PublicKey, oracleSource: any, data: OraclePriceData) => {
				// Only process updates for our target market
				if (!publicKey.equals(this.marketConfig.oracle)) return;

				if (!this.isValidOracleUpdate(data)) return;

				const oraclePrice = convertToNumber(data.price);

				if (this.shouldSkipOracleUpdate(oraclePrice)) return;

				this.handleOracleUpdate(oraclePrice).catch((error) => {
					log.error('ORACLE', 'Oracle update handler failed', error);
				});
			}
		);

		log.info('ORACLE', 'Oracle listener setup complete');
	}

	/**
	 * Validate oracle data to prevent acting on corrupt prices
	 */
	private isValidOracleUpdate(data: OraclePriceData): boolean {
		const oraclePrice = convertToNumber(data.price);

		if (oraclePrice <= 0) {
			log.warn('ORACLE', 'Invalid oracle price received', {
				price: data.price.toString(),
			});
			return false;
		}

		// Skip if price drops >99% (likely corrupt data)
		if (
			oraclePrice <= 0 ||
			(this.state.oraclePrice > 0 &&
				oraclePrice < this.state.oraclePrice * 0.01)
		) {
			log.warn('ORACLE', 'Skipping suspicious oracle price drop', {
				current: this.state.oraclePrice,
				new: oraclePrice,
			});
			return false;
		}

		return true;
	}

	/**
	 * Debounce oracle updates to avoid excessive quote changes
	 */
	private shouldSkipOracleUpdate(newPrice: number): boolean {
		const now = Date.now();

		// Debounce rapid updates
		if (now - this.state.lastOracleUpdate < MM_CONFIG.DEBOUNCE_MS) {
			return true;
		}

		// Skip if price change is too small
		if (this.state.previousOraclePrice > 0) {
			const changeBps =
				Math.abs(
					(newPrice - this.state.previousOraclePrice) /
						this.state.previousOraclePrice
				) * 10000;

			if (changeBps < MM_CONFIG.ORACLE_CHANGE_THRESHOLD_BPS) {
				return true;
			}
		}

		return false;
	}

	/**
	 * Setup fill event listener to track our order executions
	 */
	private async setupFillEventListener(): Promise<void> {
		const userAccountPublicKey =
			await this.driftClient.getUserAccountPublicKey();
		const userAccountString = userAccountPublicKey.toString();

		log.info('FILL_LISTENER', 'Setting up fill event listener', {
			userAccount: userAccountString,
		});

		this.eventSubscriber.eventEmitter.on('newEvent', (event: any) => {
			if (!this.isOurFillEvent(event, userAccountString)) return;

			const fillData = this.extractFillData(event, userAccountString);
			if (!fillData) return;

			this.handleFill(fillData);
		});
	}

	/**
	 * Check if fill event is for our orders
	 */
	private isOurFillEvent(event: any, userAccountString: string): boolean {
		return (
			event.eventType === 'OrderActionRecord' &&
			event.marketIndex === this.marketIndex &&
			event.marketType &&
			'perp' in event.marketType &&
			event.action &&
			'fill' in event.action &&
			(event.maker?.toString() === userAccountString ||
				event.taker?.toString() === userAccountString)
		);
	}

	/**
	 * Extract fill details from event data
	 */
	private extractFillData(
		event: any,
		userAccountString: string
	): {
		price: number;
		size: number;
		side: 'BUY' | 'SELL';
	} | null {
		if (!event.baseAssetAmountFilled || !event.quoteAssetAmountFilled) {
			log.warn('FILL_EXTRACT', 'Missing fill amounts in event');
			return null;
		}

		try {
			const baseAmount =
				event.baseAssetAmountFilled.toNumber() / BASE_PRECISION.toNumber();
			const quoteAmount =
				event.quoteAssetAmountFilled.toNumber() / QUOTE_PRECISION.toNumber();

			const price = quoteAmount / baseAmount;
			const size = Math.abs(baseAmount);

			// Determine if we were maker or taker, and the side
			const isMaker = event.maker?.toString() === userAccountString;
			const side = isMaker
				? event.makerOrderDirection && 'long' in event.makerOrderDirection
					? 'BUY'
					: 'SELL'
				: event.takerOrderDirection && 'long' in event.takerOrderDirection
					? 'BUY'
					: 'SELL';

			return { price, size, side };
		} catch (error) {
			log.error('FILL_EXTRACT', 'Failed to extract fill data', error as Error);
			return null;
		}
	}

	/**
	 * Handle fill execution and update P&L tracking
	 */
	private handleFill(fill: {
		side: 'BUY' | 'SELL';
		size: number;
		price: number;
	}): void {
		// Add to appropriate FIFO queue
		if (fill.side === 'BUY') {
			this.buyQueue.push({
				timestamp: Date.now(),
				size: fill.size,
				price: fill.price,
			});
		} else {
			this.sellQueue.push({
				timestamp: Date.now(),
				size: fill.size,
				price: fill.price,
			});
		}

		log.info('FILL', 'Fill detected', {
			side: fill.side,
			size: fill.size.toFixed(6),
			price: fill.price.toFixed(2),
		});

		this.processFIFOMatching();
	}

	/**
	 * FIFO matching to calculate realized P&L
	 * Matches oldest buys with oldest sells to track true profitability
	 */
	private processFIFOMatching(): void {
		while (this.buyQueue.length > 0 && this.sellQueue.length > 0) {
			const oldestBuy = this.buyQueue[0]!;
			const oldestSell = this.sellQueue[0]!;

			// Match the smaller of the two sizes
			const matchSize = Math.min(oldestBuy.size, oldestSell.size);
			const pnl = (oldestSell.price - oldestBuy.price) * matchSize;

			this.state.realizedPnl += pnl;
			this.state.totalMatches++;

			log.info('MATCH', 'Round trip completed', {
				matchNumber: this.state.totalMatches,
				matchSize: matchSize.toFixed(6),
				buyPrice: oldestBuy.price.toFixed(2),
				sellPrice: oldestSell.price.toFixed(2),
				pnl: pnl.toFixed(2),
				totalPnl: this.state.realizedPnl.toFixed(2),
			});

			// Update remaining sizes
			oldestBuy.size -= matchSize;
			oldestSell.size -= matchSize;

			// Remove fully consumed trades
			if (oldestBuy.size <= 0.000001) {
				this.buyQueue.shift();
			}
			if (oldestSell.size <= 0.000001) {
				this.sellQueue.shift();
			}
		}
	}

	/**
	 * Main oracle update handler - triggers quote recalculation
	 */
	private async handleOracleUpdate(newOraclePrice: number): Promise<void> {
		if (!this.state.isRunning) return;

		// Skip if already processing
		if (this.state.isProcessing) {
			return;
		}

		const now = Date.now();

		log.info('ORACLE', 'Processing oracle update', {
			newPrice: newOraclePrice.toFixed(2),
			oldPrice: this.state.oraclePrice.toFixed(2),
		});

		// Update price state
		this.state.previousOraclePrice = this.state.oraclePrice;
		this.state.oraclePrice = newOraclePrice;
		this.state.lastOracleUpdate = now;

		try {
			await this.updateQuotes();
		} catch (error) {
			log.error('UPDATE', 'Failed to update quotes', error as Error);
		}
	}

	/**
	 * Calculate new quotes and update orders
	 */
	private async updateQuotes(): Promise<void> {
		this.state.isProcessing = true;

		try {
			const currentPosition = this.getCurrentPosition();
			const quotes = this.calculateQuotes(currentPosition);

			if (!this.validateQuotes(quotes)) {
				log.warn('QUOTES', 'Invalid quotes generated, skipping update');
				return;
			}

			await this.updateOrders(quotes);
		} finally {
			this.state.isProcessing = false;
		}
	}
	/**
	 * Get current position from Drift client
	 */
	private getCurrentPosition(): number {
		try {
			const perpPosition = this.driftClient
				.getUser()
				.getPerpPosition(this.marketIndex);
			return perpPosition
				? perpPosition.baseAssetAmount.toNumber() / BASE_PRECISION.toNumber()
				: 0;
		} catch (error) {
			log.warn('POSITION', 'Failed to get position', error as Error);
			return 0;
		}
	}

	/**
	 * Calculate bid/ask quotes with inventory skewing
	 * Core market making logic: base spread + position-based skew
	 */
	private calculateQuotes(currentPosition: number): {
		bidOffsetBps: number;
		askOffsetBps: number;
		bidSize: number;
		askSize: number;
	} {
		const baseSpreadBps = MM_CONFIG.BASE_SPREAD_BPS;
		const halfSpreadBps = baseSpreadBps / 2;

		// Calculate position ratio and skew
		const positionRatio = currentPosition / MM_CONFIG.MAX_POSITION;
		const skewBps = Math.abs(positionRatio) * MM_CONFIG.MAX_SKEW_BPS;

		let bidOffsetBps = halfSpreadBps;
		let askOffsetBps = halfSpreadBps;

		// Inventory skewing: widen quotes away from position direction
		if (currentPosition > 0) {
			// Long position: widen bids, tighten asks to encourage selling
			bidOffsetBps += skewBps;
			askOffsetBps = askOffsetBps - skewBps;
		} else if (currentPosition < 0) {
			// Short position: tighten bids, widen asks to encourage buying
			bidOffsetBps = bidOffsetBps - skewBps;
			askOffsetBps += skewBps;
		}

		return {
			bidOffsetBps,
			askOffsetBps,
			bidSize: MM_CONFIG.ORDER_SIZE,
			askSize: MM_CONFIG.ORDER_SIZE,
		};
	}

	/**
	 * Validate quotes before placing orders
	 */
	private validateQuotes(quotes: {
		bidOffsetBps: number;
		askOffsetBps: number;
		bidSize: number;
		askSize: number;
	}): boolean {
		if (quotes.bidOffsetBps <= 0 || quotes.askOffsetBps <= 0) {
			log.warn('VALIDATION', 'Invalid offsets in quotes');
			return false;
		}

		if (quotes.bidSize <= 0 || quotes.askSize <= 0) {
			log.warn('VALIDATION', 'Invalid sizes in quotes');
			return false;
		}

		return true;
	}

	/**
	 * Update orders: cancel existing + place new quotes
	 */
	private async updateOrders(quotes: {
		bidOffsetBps: number;
		askOffsetBps: number;
		bidSize: number;
		askSize: number;
	}): Promise<void> {
		try {
			if (this.state.hasActiveOrders) {
				await this.driftClient.cancelOrders(MarketType.PERP, this.marketIndex);
			}

			// Convert BPS to price offsets
			const bidPriceOffset =
				-this.state.oraclePrice * (quotes.bidOffsetBps / 10000);
			const askPriceOffset =
				this.state.oraclePrice * (quotes.askOffsetBps / 10000);

			const bidOrderParams = {
				orderType: OrderType.LIMIT,
				marketIndex: this.marketIndex,
				direction: PositionDirection.LONG,
				baseAssetAmount: this.driftClient.convertToPerpPrecision(
					quotes.bidSize
				),
				price: new BN(0),
				oraclePriceOffset: Math.round(
					bidPriceOffset * QUOTE_PRECISION.toNumber()
				),
				postOnly: PostOnlyParams.TRY_POST_ONLY,
				marketType: MarketType.PERP,
			};

			const askOrderParams = {
				orderType: OrderType.LIMIT,
				marketIndex: this.marketIndex,
				direction: PositionDirection.SHORT,
				baseAssetAmount: this.driftClient.convertToPerpPrecision(
					quotes.askSize
				),
				price: new BN(0),
				oraclePriceOffset: Math.round(
					askPriceOffset * QUOTE_PRECISION.toNumber()
				),
				postOnly: PostOnlyParams.TRY_POST_ONLY,
				marketType: MarketType.PERP,
			};

			await this.driftClient.placePerpOrder(bidOrderParams);
			await this.driftClient.placePerpOrder(askOrderParams);
			this.state.hasActiveOrders = true;

			log.info('ORDERS', 'Oracle orders updated successfully', {
				bidOffset: bidPriceOffset.toFixed(2),
				askOffset: askPriceOffset.toFixed(2),
			});
		} catch (error) {
			log.error('ORDERS', 'Failed to update oracle orders', error as Error);
			this.state.hasActiveOrders = false;
		}
	}

	/**
	 * Close current position with market order
	 */
	private async closeCurrentPosition(): Promise<void> {
		const currentPosition = this.getCurrentPosition();

		if (Math.abs(currentPosition) < 0.0001) {
			return; // No position to close
		}

		try {
			const closeDirection =
				currentPosition > 0 ? PositionDirection.SHORT : PositionDirection.LONG;

			const closeOrderParams = {
				orderType: OrderType.MARKET,
				marketIndex: this.marketIndex,
				direction: closeDirection,
				baseAssetAmount: this.driftClient.convertToPerpPrecision(
					Math.abs(currentPosition)
				),
				reduceOnly: true,
				marketType: MarketType.PERP,
			};

			await this.driftClient.placePerpOrder(closeOrderParams);

			log.info('CLOSE', 'Position closed', {
				size: currentPosition.toFixed(6),
				direction: closeDirection === PositionDirection.LONG ? 'BUY' : 'SELL',
			});
		} catch (error) {
			log.error('CLOSE', 'Failed to close position', error as Error);
		}
	}

	/**
	 * Start the market maker
	 */
	start(): void {
		this.state.isRunning = true;
		log.info('START', 'Example Maker started');
	}

	/**
	 * Stop the market maker and clean up
	 */
	async stop(): Promise<void> {
		log.info('STOP', 'Stopping Example Maker');
		this.state.isRunning = false;

		// Cancel any active orders
		try {
			if (this.state.hasActiveOrders) {
				await this.driftClient.cancelOrders(MarketType.PERP, this.marketIndex);
			}

			// Close existing positions
			await this.closeCurrentPosition();
		} catch (error) {
			log.error(
				'STOP',
				'Failed to cancel orders during shutdown',
				error as Error
			);
		}

		// Unsubscribe from all data feeds
		try {
			await this.dlobSubscriber?.unsubscribe();
			await this.slotSubscriber?.unsubscribe();
			await this.orderSubscriber?.unsubscribe();
			await this.eventSubscriber?.unsubscribe();
			await this.driftClient?.unsubscribe();
		} catch (error) {
			log.error('STOP', 'Failed to unsubscribe components', error as Error);
		}

		log.info('STOP', 'Example Maker stopped');
	}
}

File: rust/bots/simple-oracle-limit-maker/src/maker.rs
use anyhow::Result;
use drift_rs::{
    types::{
        Context, MarketId, MarketType, OrderParams, OrderType, PerpPosition, PositionDirection,
        PostOnlyParam,
    },
    DriftClient, GrpcSubscribeOpts, Pubkey, RpcClient, Wallet,
};
use log::{debug, error, info};
use solana_sdk::commitment_config::CommitmentLevel;
use std::{
    env,
    str::FromStr,
    time::{SystemTime, UNIX_EPOCH},
};
use tokio::time::Duration;

// Drift precision constants
const BASE_PRECISION: f64 = 1_000_000_000.0; // 1e9
const QUOTE_PRECISION: f64 = 1_000_000.0; // 1e6

/// Bot configuration parameters
#[derive(Debug, Clone)]
pub struct BotConfig {
    pub target_market: String,            // Market
    pub order_size: f64,                  // Amount per order (side)
    pub max_position: f64,                // Maximum position size before skewing
    pub base_spread_bps: u16,             // Minimum spread around oracle price
    pub max_skew_bps: u16,                // Maximum additional skew when at max position
    pub debounce_ms: u64,                 // Minimum time between oracle updates
    pub oracle_change_threshold_bps: u16, // Minimum price change to trigger update
    pub authority: Option<String>,        // Authority pubkey (for delegation)
    pub subaccount_id: u16,               // Subaccount ID
}

#[derive(Debug)]
struct QuoteParams {
    bid_offset_bps: f64,
    ask_offset_bps: f64,
    size: f64,
}

/// Market Maker bot
pub struct OracleLimitMakerBot {
    client: DriftClient,
    config: BotConfig,
    // Market identification
    market_id: MarketId,
    // Bot state
    oracle_price: i64,
    prev_oracle_price: i64,
    last_oracle_update: u64,
    has_active_orders: bool,
    is_running: bool,
    is_processing: bool,
}

impl OracleLimitMakerBot {
    pub async fn new(config: BotConfig) -> Result<Self> {
        info!("Initializing oracle limit maker bot...");

        let client = Self::init_drift_client().await?;

        // Get market_id directly
        let market_id = client
            .market_lookup(&config.target_market)
            .ok_or_else(|| anyhow::anyhow!("Market symbol '{}' not found", config.target_market))?;

        info!("Found market: {} -> {:?}", config.target_market, market_id);

        // Subscribe to oracle feed
        // Comment out if using gRPC
        // client.subscribe_oracles(&[market_id]).await?;
        // info!(
        //     "Subscribed to oracle feed for market: {}",
        //     config.target_market
        // );

        // Get initial oracle price
        let initial_oracle_price = client.oracle_price(market_id).await.unwrap_or(0);
        info!(
            "Initial oracle price: ${:.2}",
            initial_oracle_price as f64 / QUOTE_PRECISION
        );

        info!(
            "Bot initialized for market: {} (subaccount: {})",
            config.target_market, config.subaccount_id
        );

        Ok(Self {
            client,
            config,
            market_id,
            oracle_price: initial_oracle_price,
            prev_oracle_price: 0,
            last_oracle_update: 0,
            has_active_orders: false,
            is_running: false,
            is_processing: false,
        })
    }

    async fn init_drift_client() -> Result<DriftClient> {
        let rpc_endpoint = env::var("RPC_ENDPOINT").expect("RPC_ENDPOINT not set");
        let private_key = env::var("PRIVATE_KEY").expect("PRIVATE_KEY not set");

        let context = Context::MainNet;
        let wallet = Wallet::try_from_str(&private_key).unwrap();
        let rpc_client = RpcClient::new(rpc_endpoint);

        let client = DriftClient::new(context, rpc_client, wallet)
            .await
            .expect("Failed to initialize client");

        // gRPC version
        let grpc_url = env::var("GRPC_URL").expect("GRPC_URL not set");
        let grpc_token = env::var("GRPC_X_TOKEN").expect("GRPC_X_TOKEN not set");

        client
            .grpc_subscribe(
                grpc_url,
                grpc_token,
                GrpcSubscribeOpts::default().commitment(CommitmentLevel::Processed),
                true,
            )
            .await?;

        info!(
            "Connected to Drift with wallet: {}",
            client.wallet().authority()
        );
        Ok(client)
    }

    /// Start the main trading loop
    pub async fn start(&mut self) -> Result<()> {
        self.is_running = true;
        info!("Starting oracle limit maker...");

        while self.is_running {
            if let Err(e) = self.process_cycle().await {
                error!("Trading cycle failed: {}", e);
                tokio::time::sleep(Duration::from_secs(5)).await; // Back off on errors
                continue;
            }

            tokio::time::sleep(Duration::from_millis(self.config.debounce_ms)).await;
        }

        Ok(())
    }

    /// Process single trading cycle
    async fn process_cycle(&mut self) -> Result<()> {
        if self.is_processing {
            return Ok(());
        }

        // Get current oracle price
        let current_price = self.client.oracle_price(self.market_id).await?;

        // Check if price changed significantly
        if self.should_update_quotes(current_price) {
            self.handle_oracle_update(current_price).await?;
        } else {
            debug!(
                "Oracle price unchanged or within threshold: ${:.2}",
                current_price as f64 / QUOTE_PRECISION
            );
        }

        Ok(())
    }

    /// Check if we should update quotes based on price change and debounce
    fn should_update_quotes(&self, new_price: i64) -> bool {
        let now = get_current_timestamp();

        // Debounce rapid updates
        if now - self.last_oracle_update < self.config.debounce_ms {
            return false;
        }

        // Skip if price change is too small
        if self.prev_oracle_price > 0 {
            let change_bps = ((new_price - self.prev_oracle_price).abs() as f64
                / self.prev_oracle_price as f64)
                * 10000.0;

            if change_bps < self.config.oracle_change_threshold_bps as f64 {
                return false;
            }

            debug!("Oracle price change: {:.2} bps", change_bps);
        }

        true
    }

    /// Handle oracle price update
    async fn handle_oracle_update(&mut self, new_oracle_price: i64) -> Result<()> {
        if self.is_processing {
            return Ok(());
        }
        self.is_processing = true;

        info!(
            "Processing oracle update: ${:.2} -> ${:.2}",
            self.oracle_price as f64 / QUOTE_PRECISION,
            new_oracle_price as f64 / QUOTE_PRECISION
        );

        // Update price state
        self.prev_oracle_price = self.oracle_price;
        self.oracle_price = new_oracle_price;
        self.last_oracle_update = get_current_timestamp();

        // Get current position for inventory skewing
        let current_position = self.get_current_position().await?;
        let position_size =
            current_position.map(|p| p.base_asset_amount).unwrap_or(0) as f64 / BASE_PRECISION;

        info!(
            "Current position: {:.6} {}",
            position_size,
            self.config.target_market.replace("-PERP", "")
        );

        // Calculate quotes with inventory skewing
        let quotes = self.calculate_quotes(position_size);

        // Update orders: cancel existing + place new
        self.update_orders(quotes).await?;

        self.is_processing = false;
        Ok(())
    }

    /// Get current perp position
    async fn get_current_position(&self) -> Result<Option<PerpPosition>> {
        let subaccount = self.get_subaccount();
        let user_account = self.client.get_user_account(&subaccount).await?;

        Ok(user_account
            .perp_positions
            .iter()
            .find(|pos| pos.market_index == self.market_id.index())
            .cloned())
    }

    /// Calculate bid/ask quotes with inventory skewing
    fn calculate_quotes(&self, current_position: f64) -> QuoteParams {
        let base_spread_bps = self.config.base_spread_bps as f64;
        let half_spread_bps = base_spread_bps / 2.0;

        // Calculate position ratio and skew
        let position_ratio = current_position / self.config.max_position;
        let skew_bps = position_ratio.abs() * self.config.max_skew_bps as f64;

        let mut bid_offset_bps = half_spread_bps;
        let mut ask_offset_bps = half_spread_bps;

        // Inventory skewing: widen quotes away from position direction
        if current_position > 0.0 {
            // Long position: widen bids, tighten asks to encourage selling
            bid_offset_bps += skew_bps;
            ask_offset_bps = (ask_offset_bps - skew_bps).max(1.0); // Ensure positive
            debug!(
                "Long position detected, widening bids (+{:.1} bps), tightening asks (-{:.1} bps)",
                skew_bps, skew_bps
            );
        } else if current_position < 0.0 {
            // Short position: tighten bids, widen asks to encourage buying
            bid_offset_bps = (bid_offset_bps - skew_bps).max(1.0); // Ensure positive
            ask_offset_bps += skew_bps;
            debug!(
                "Short position detected, tightening bids (-{:.1} bps), widening asks (+{:.1} bps)",
                skew_bps, skew_bps
            );
        } else {
            debug!("No position, using base spread");
        }

        QuoteParams {
            bid_offset_bps,
            ask_offset_bps,
            size: self.config.order_size,
        }
    }

    /// Update orders: cancel existing and place new quotes
    async fn update_orders(&mut self, quotes: QuoteParams) -> Result<()> {
        let subaccount = self.get_subaccount();

        // Calculate oracle price offsets
        let oracle_price_f64 = self.oracle_price as f64;
        let bid_price_offset = -(oracle_price_f64 * quotes.bid_offset_bps / 10000.0) as i32;
        let ask_price_offset = (oracle_price_f64 * quotes.ask_offset_bps / 10000.0) as i32;

        // For logging, calculate display prices
        let bid_display_price = oracle_price_f64 + (bid_price_offset as f64);
        let ask_display_price = oracle_price_f64 + (ask_price_offset as f64);

        // Create bid order (buy) with oracle offset
        let bid_order = OrderParams {
            order_type: OrderType::Limit,
            market_type: MarketType::Perp,
            direction: PositionDirection::Long,
            base_asset_amount: (quotes.size * BASE_PRECISION) as u64,
            market_index: self.market_id.index(),
            price: 0,                                    // Set to 0 when using oracle offset
            oracle_price_offset: Some(bid_price_offset), // Negative for bid
            post_only: PostOnlyParam::TryPostOnly,
            ..Default::default()
        };

        // Create ask order (sell) with oracle offset
        let ask_order = OrderParams {
            order_type: OrderType::Limit,
            market_type: MarketType::Perp,
            direction: PositionDirection::Short,
            base_asset_amount: (quotes.size * BASE_PRECISION) as u64,
            market_index: self.market_id.index(),
            price: 0,                                    // Set to 0 when using oracle offset
            oracle_price_offset: Some(ask_price_offset), // Positive for ask
            post_only: PostOnlyParam::TryPostOnly,
            ..Default::default()
        };

        // Single atomic transaction: cancel + place both orders
        let cancel_and_place_tx = self
            .client
            .init_tx(&subaccount, self.is_delegated())
            .await?
            .cancel_orders((self.market_id.index(), MarketType::Perp), None)
            .place_orders(vec![bid_order, ask_order])
            .build();

        let signature = self.client.sign_and_send(cancel_and_place_tx).await?;
        self.has_active_orders = true;

        info!(
            "Updated quotes - Bid: ${:.2} (-{:.1} bps), Ask: ${:.2} (+{:.1} bps), Size: {:.6}, Sig: {}",
            bid_display_price / QUOTE_PRECISION,
            quotes.bid_offset_bps,
            ask_display_price / QUOTE_PRECISION,
            quotes.ask_offset_bps,
            quotes.size,
            signature
        );

        Ok(())
    }
    /// Get subaccount pubkey
    fn get_subaccount(&self) -> Pubkey {
        match &self.config.authority {
            Some(authority_str) => {
                let authority = Pubkey::from_str(authority_str).expect("Invalid authority pubkey");
                Wallet::derive_user_account(&authority, self.config.subaccount_id)
            }
            None => self.client.wallet().sub_account(self.config.subaccount_id),
        }
    }

    /// Check if using delegated signing
    fn is_delegated(&self) -> bool {
        self.config.authority.is_some()
    }

    /// Stop the bot and clean up
    pub async fn stop(&mut self) {
        info!("Stopping oracle limit maker bot...");
        self.is_running = false;

        // Cancel any active orders
        if self.has_active_orders {
            info!("Cancelling active orders before shutdown");
            let subaccount = self.get_subaccount();
            if let Ok(cancel_tx) = self
                .client
                .init_tx(&subaccount, self.is_delegated())
                .await
                .map(|tx| tx.cancel_all_orders().build())
            {
                if let Err(e) = self.client.sign_and_send(cancel_tx).await {
                    error!("Failed to cancel orders during shutdown: {}", e);
                } else {
                    info!("Successfully cancelled orders");
                }
            }
        }

        // Close any open positions
        if let Ok(current_position) = self.get_current_position().await {
            if let Some(pos) = current_position {
                if pos.base_asset_amount != 0 {
                    info!("Closing open position before shutdown");
                    let close_direction = if pos.base_asset_amount > 0 {
                        PositionDirection::Short
                    } else {
                        PositionDirection::Long
                    };

                    let close_order = OrderParams {
                        order_type: OrderType::Market,
                        market_type: MarketType::Perp,
                        direction: close_direction,
                        base_asset_amount: pos.base_asset_amount.unsigned_abs(),
                        market_index: self.market_id.index(),
                        reduce_only: true,
                        ..Default::default()
                    };

                    if let Ok(close_tx) = self
                        .client
                        .init_tx(&self.get_subaccount(), self.is_delegated())
                        .await
                        .map(|tx| tx.place_orders(vec![close_order]).build())
                    {
                        if let Err(e) = self.client.sign_and_send(close_tx).await {
                            error!("Failed to close position during shutdown: {}", e);
                        } else {
                            info!("Successfully closed position");
                        }
                    }
                }
            }
        }

        // Unsubscribe from oracle feed
        if let Err(e) = self.client.unsubscribe().await {
            error!("Failed to unsubscribe: {}", e);
        } else {
            info!("Unsubscribed from oracle feeds");
        }

        info!("Oracle limit maker bot stopped");
    }
}

/// Get current timestamp in milliseconds
fn get_current_timestamp() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as u64
}
